DirectX11Proj/IApplication.cpp

```diff
@@ -2,47 +2,46 @@
 
 #include <iostream>
 
+#include "easylogging++.h"
 
 #include "IScene.h"
 #include "systemclass.h"
 
 
-IApplication::IApplication() : mpCurrentScene(nullptr), mShouldQuit(false)
+
+IApplication::IApplication() : mShouldQuit(false)
 {
 }
 
 
 IApplication::~IApplication()
 {
-	if (mpCurrentScene)
-	{
-		// fucking evil (wow). should really fix something about this (and think of a decent scene management system)
-		delete mpCurrentScene;
-	}
+
 }
 
 // Setting the current scene also initializes it
-void IApplication::LoadScene(IScene* aScene)
+void IApplication::LoadScene(std::unique_ptr<IScene> apScene)
 {
-	assert(aScene != nullptr);
+	assert(apScene != nullptr);
 
-	if (aScene == nullptr)
+	if (apScene == nullptr)
 	{
 		std::cout << "Can not set a nullptr as current scene!" << std::endl;
 		return;
 	}
 
+	// Delete scene last used.
 	if (mpCurrentScene != nullptr)
 	{
-		std::cout << "There is already a scene loaded!" << std::endl;
-		return;
+		mpCurrentScene.release();
 	}
 
-	this->mpCurrentScene = aScene;
+	mpCurrentScene = std::move(apScene);;
+
 	if (!mpCurrentScene->HasBeenInitialized())
 	{
-		aScene->Init();
-		aScene->mInitialized = true;
+		mpCurrentScene->Init();
+		mpCurrentScene->mInitialized = true;
```

Comment:
It's a bit hard because I do not want to put this burden on the user, unless it calls the virtual function as well when I call the overidden function? I do not know